### PR TITLE
API and migration reference from .NET 3.1 / .NET 5 to .NET 6

### PR DESCRIPTION
**Pull request by vlada-shubina**
The PR adds "Inside the Template Engine" doc that explains the how template engine is built and its APIs.
Also adds migration reference from .NET 3.1 / .NET 5 to .NET 6

### CODE REVIEW

**Comment by vlada-shubina** [comment]
@DavidKarlas @bekir-ozturk  let's merge this and post to Wiki

File: docs/api/Inside-the-Template-Engine.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
+# Inside the Template Engine
+
+The template engine is composed of several different subsystems, which are
+designed to separate gathering templates, instantiating template and processing
+templates.
+
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:6]: A "table of contents" would be useful here.

File: docs/api/Inside-the-Template-Engine.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+Those properties are used to identify host to various built-in components
+explained below.
+
+If you are considering using the template engine core, you need to make
+implementation of this interface representing your host. Template engine
+provides default implementation:
+[DefaultTemplateEngineHost](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Edge/DefaultTemplateEngineHost.cs).
+
+#### IPhysicalFileSystem
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:44]: Missing link.
Is the header depth correct?

File: docs/api/Inside-the-Template-Engine.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+Abstraction over file system, accessible from ITemplateEngineHost.
+
+Each template engine host should support physical and in-memory file system and
+virtualize file system under given path. Switch between physical and in-memory
+is done via ITemplateEngineHost.VirtualizePath method.
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:50]:
```suggestion
Each template engine host should support physical and in-memory file system and
virtualize the file system under a given path. Switching between physical and in-memory
file systems is done via `ITemplateEngineHost.VirtualizePath` method.
```

File: docs/api/Inside-the-Template-Engine.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+Template engine provides default implementation of file systems:
+[PhysicalFileSystem](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Utils/PhysicalFileSystem.cs)
+and
+[InMemoryFileSystem](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Utils/InMemoryFileSystem.cs).
+
+### [IComponentManager](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IComponentManager.cs)
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:57]: Maybe we should explain what a component is before we explain component manager

File: docs/api/Inside-the-Template-Engine.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+ComponentManager is responsible for loading components provided by Host from
+ITemplateEngineHost.BuiltIns, but additional components can be added via
+IComponentManager.AddComponent or they are dynamically loaded from
+TemplatePackage during scanning similar to templates. ComponentManager
+implementation is not publicly accessible, however it is used when default
+implementation of IEngineEnvironmentSettings is used.
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:64]:
```suggestion
`ComponentManager` is responsible for loading components provided by the host from
`ITemplateEngineHost.BuiltIns`. Additional components can be added via
`IComponentManager.AddComponent`. Similar to templates, components are also 
dynamically loaded from `TemplatePackage`s during scanning. `ComponentManager`
implementation is not publicly accessible; however, it is used when default
implementation of `IEngineEnvironmentSettings` is used.
```

File: docs/api/Inside-the-Template-Engine.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+### [IEnvironment](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IEnvironment.cs)
+
+Abstraction over environment variables. Default implementation uses system
+environment variables.
+
+### [IPathInfo](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IPathInfo.cs)
+
+Provides the main paths used by template engine: global settings path, host
+settings path host version settings path.
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:74]:
```suggestion
Provides the main paths used by template engine: global settings path, host
settings path, host version settings path.
```

File: docs/api/Inside-the-Template-Engine.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+TemplatePackage during scanning similar to templates. ComponentManager
+implementation is not publicly accessible, however it is used when default
+implementation of IEngineEnvironmentSettings is used.
+
+### [IEnvironment](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IEnvironment.cs)
+
+Abstraction over environment variables. Default implementation uses system
+environment variables.
+
+### [IPathInfo](https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Abstractions/IPathInfo.cs)
```
Comment (bekir-ozturk) [inline docs/api/Inside-the-Template-Engine.md:71]: Could these be listed at the very end? These look unimportant in understanding the main architecture. They are just simple abstractions the user can figure out while coding. But they make the document hard to follow for one reason: we have too many headers with various depths.

**Review by bekir-ozturk — CHANGES_REQUESTED** [review]
I think we have too many headers at different depths to understand the hierarchy of this doc. If possible, let's only keep the main subsystems and mention the rest in paragraphs. Or create a separate page for each subsystem.

File: docs/api/Inside-the-Template-Engine.md (hunk 7 above)
Comment (vlada-shubina) [reply docs/api/Inside-the-Template-Engine.md:70]: They belong to environment settings not sure, if it is a good trade-off

File: docs/api/Inside-the-Template-Engine.md (hunk 4 above)
Comment (vlada-shubina) [reply docs/api/Inside-the-Template-Engine.md:56]: Added the link to components section below

File: docs/api/Inside-the-Template-Engine.md (hunk 2 above)
Comment (vlada-shubina) [reply docs/api/Inside-the-Template-Engine.md:43]: Added link, depth is correct

**Review by bekir-ozturk — APPROVED** [review]
I still think we could get this a bit easier to follow, but it's good enough to merge.
